webview-ui/src/components/chat/VoiceRecorder.tsx

```diff
@@ -0,0 +1,271 @@
+			setError(null)
+			onProcessingStateChange?.(false)
+			onTranscription("")
+
+			// Call Extension Host to cancel recording
+			const response = await DictationServiceClient.cancelRecording(EmptyRequest.create({}))
+
+			if (!response.success) {
+				console.error("Failed to cancel recording:", response.error)
+				setError(response.error || "Failed to cancel recording")
+				return
+			}
+
+			console.log("Recording canceled successfully")
+		} catch (error) {
+			console.error("Error canceling recording:", error)
+			const errorMessage = error instanceof Error ? error.message : "Failed to cancel recording"
+			setError(errorMessage)
+		}
+	}, [onProcessingStateChange, onTranscription])
+
+	const handleStartClick = useCallback(() => {
+		if (disabled || isProcessing || isStarting) {
+			return
+		}
+		if (error) {
+			return setError(null)
+		}
+		startRecording()
+	}, [startRecording, disabled, isProcessing, isStarting, error])
+
+	const handleCancelClick = useCallback(() => {
+		if (disabled || isProcessing) {
+			return
+		}
+		cancelRecording()
+	}, [cancelRecording, disabled, isProcessing])
+
+	const handleStopClick = useCallback(() => {
+		if (disabled || isProcessing) {
+			return
+		}
+		stopRecording()
+	}, [stopRecording, disabled, isProcessing])
+
+	const iconAnimation = isProcessing || isStarting ? "animate-spin" : ""
+	const iconAdjustment = isProcessing || isStarting ? "mt-0" : error ? "mt-1" : "mt-0.5"
+	// When not recording, show single mic button
+	if (!isRecording) {
+		const iconClass = isProcessing
+			? "codicon-loading"
+			: isStarting
+				? "codicon-loading"
+				: error
+					? "codicon-error"
+					: "codicon-mic"
+		const iconColor = error ? "text-error" : ""
+		const tooltipContent = isProcessing
+			? "Transcribing..."
+			: isStarting
+				? "Starting recording..."
+				: error
+					? `Error: ${error}`
+					: null
+
+		return (
+			<HeroTooltip content={tooltipContent} placement="top">
+				<div
+					className={`input-icon-button mr-1.5 text-base ${iconAdjustment} ${iconAnimation} ${disabled || isProcessing || isStarting ? "disabled" : ""}`}
+					onClick={handleStartClick}
+					style={{ color: iconColor }}>
+					<span className={`codicon ${iconClass}`} />
+				</div>
+			</HeroTooltip>
+		)
+	}
+
+	return (
+		<div className={`flex items-center ${isRecording ? "mr-0.5" : "mr-1.5"}`}>
+			<HeroTooltip
+				content={`Stop Recording (${formatSeconds(recordingDuration)}/${formatSeconds(MAX_DURATION)})`}
+				placement="top">
+				<div
+					className={cn("input-icon-button mr-1.5 text-base", iconAdjustment, iconAnimation, {
+						disabled: disabled || isProcessing,
+					})}
+					onClick={handleStopClick}>
+					<span className="codicon codicon-stop-circle" />
+				</div>
+			</HeroTooltip>
+			<HeroTooltip content="Cancel Recording" placement="top">
+				<div
+					className={`input-icon-button text-base mt-1 text-[var(--vscode-textForeground)] ${disabled || isProcessing ? "disabled" : ""}`}
```

Comment:
is vscode-textForeground valid? Alternative you can use `text-foreground` as `vscode-foreground`
```suggestion
					className={`input-icon-button text-base mt-1 text-foreground ${disabled || isProcessing ? "disabled" : ""}`}
```